File: implementation/swift/GameEventScriptSyntaxHighlighter/Sources/GameEventScriptSyntaxHighlighter/HighlightingModel.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

/// The bundled source or assembler grammar used for presentation.
public enum GameEventScriptSyntaxLanguage: String, Sendable {
    /// Game Event Script source, including incomplete editor input.
    case ges
    /// GESA assembler with embedded GES source regions.
    case gesa
}

/// Theme-independent categories derived from TextMate scopes; these are not compiler tokens.
public enum GameEventScriptSyntaxKind: String, CaseIterable, Sendable {
    /// Text without a more specific category, including whitespace.
    case plain
    /// A keyword, word operator or boolean literal.
    case keyword
    /// A built-in function or collection selector.
    case builtin
    /// A variable or parameter identifier.
    case identifier
    /// A message or handler name.
    case message
    /// A source type name.
    case type
    /// A tag name.
    case tag
    /// A named constant.
    case constant
    /// A numeric literal, including units or dice syntax.
    case number
    /// A string literal, delimiter or escape.
    case string
    /// A comment.
    case comment
    /// Punctuation such as parentheses and separators.
    case symbol
    /// A declared function or predicate name.
    case function
    /// A module name.
    case module
    /// An assembler label or symbol reference.
    case label
    /// An assembler register.
    case register
}

/// An immutable UTF-16 range with its category and outer-to-inner TextMate scope stack.
public struct GameEventScriptHighlightSpan: Equatable, Sendable {
    /// Zero-based UTF-16 offset in the supplied document or line.
    public let start: Int
    /// Range length in UTF-16 code units.
    public let length: Int
    /// The most specific recognized category.
    public let kind: GameEventScriptSyntaxKind
    /// Outer-to-inner TextMate scopes, including the root language scope.
    public let scopes: [String]

    init(_ start: Int, _ length: Int, _ scopes: [String]) {
        self.start = start
        self.length = length
        self.scopes = scopes
        kind = ScopeKinds.classify(scopes)
    }
}

/// Immutable, equatable line-boundary context for incremental convergence; not a serializable document or compiler state.
public struct GameEventScriptHighlightState: Equatable, Sendable {
    /// The grammar this state belongs to. States cannot be reused for another language.
    public let language: GameEventScriptSyntaxLanguage
    let frames: [Int]
}

/// A highlighting result. Failed resource bounds return no spans or usable continuation state.
public struct GameEventScriptHighlightResult: Sendable {
    /// Ordered, nonoverlapping ranges covering all input when successful.
    public let spans: [GameEventScriptHighlightSpan]
    /// Whether all input was highlighted within the resource limits.
    public var isComplete: Bool { nextState != nil }
    /// Context after the final line, or nil when incomplete. Do not cache an incomplete result.
    public let nextState: GameEventScriptHighlightState?
}

/// Invalid API input; malformed GES/GESA code is accepted and does not produce these errors.
public enum GameEventScriptHighlightingError: Error {
    /// The input-length bound must be positive.
    case invalidInputLimit
    /// A line API call contained an interior CR or LF.
    case expectedSingleLine
    /// The continuation state belongs to another language.
    case incompatibleState
    /// Rendering spans overlap, exceed the supplied text or split a Unicode scalar.
    case invalidSpans
    /// ANSI colors must be nil or one of 30...37, 39, 90...97.
    case invalidAnsiColor
}

enum ScopeKinds {
    static func classify(_ scopes: [String]) -> GameEventScriptSyntaxKind {
        for scope in scopes.reversed() {
            if scope.hasPrefix("comment") { return .comment }
            if scope.hasPrefix("string") || scope.hasPrefix("constant.character.escape") { return .string }
            if scope.hasPrefix("constant.numeric") || scope.hasPrefix("constant.language.numeric") { return .number }
            if scope.hasPrefix("variable.other.constant") { return .constant }
            if scope.hasPrefix("variable.other.register") { return .register }
            if scope.hasPrefix("variable.other.label") || scope.hasPrefix("constant.other.symbol") { return .label }
            if scope.hasPrefix("entity.name.type.class.message") || scope.hasPrefix("entity.name.function.handler") { return .message }
            if scope.hasPrefix("support.type") || scope.hasPrefix("entity.name.type") { return .type }
            if scope.hasPrefix("entity.name.namespace") { return .module }
            if scope.hasPrefix("entity.name.function") { return .function }
            if scope.hasPrefix("entity.other.attribute-name.tag") { return .tag }
            if scope.hasPrefix("support") || scope.hasPrefix("entity") { return .builtin }
            if scope.hasPrefix("keyword") || scope.hasPrefix("constant.language.boolean") { return .keyword }
            if scope.hasPrefix("variable") { return .identifier }
            if scope.hasPrefix("punctuation") { return .symbol }
        }
        return .plain
    }
}

File: implementation/csharp/GameEventScript.SyntaxHighlighter/src/HighlightingModel.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Linq;

namespace GameEventScript.SyntaxHighlighter;

/// <summary>The bundled source or assembler grammar used for presentation.</summary>
public enum GameEventScriptSyntaxLanguage
{
    /// <summary>Game Event Script source, including incomplete editor input.</summary>
    Ges,
    /// <summary>GESA assembler with embedded GES source regions.</summary>
    Gesa
}

/// <summary>Theme-independent categories derived from TextMate scopes; these are not compiler tokens.</summary>
public enum GameEventScriptSyntaxKind
{
    /// <summary>Text without a more specific category, including whitespace.</summary>
    Plain,
    /// <summary>A keyword, word operator or boolean literal.</summary>
    Keyword,
    /// <summary>A built-in function or collection selector.</summary>
    Builtin,
    /// <summary>A variable or parameter identifier.</summary>
    Identifier,
    /// <summary>A message or handler name.</summary>
    Message,
    /// <summary>A source type name.</summary>
    Type,
    /// <summary>A tag name.</summary>
    Tag,
    /// <summary>A named constant.</summary>
    Constant,
    /// <summary>A numeric literal, including units or dice syntax.</summary>
    Number,
    /// <summary>A string literal, delimiter or escape.</summary>
    String,
    /// <summary>A comment.</summary>
    Comment,
    /// <summary>Punctuation such as parentheses and separators.</summary>
    Symbol,
    /// <summary>A declared function or predicate name.</summary>
    Function,
    /// <summary>A module name.</summary>
    Module,
    /// <summary>An assembler label or symbol reference.</summary>
    Label,
    /// <summary>An assembler register.</summary>
    Register
}

/// <summary>An immutable UTF-16 range with its category and outer-to-inner TextMate scope stack.</summary>
public sealed class GameEventScriptHighlightSpan
{
    /// <summary>Zero-based UTF-16 offset in the supplied document or line.</summary>
    public int Start { get; }
    /// <summary>Range length in UTF-16 code units.</summary>
    public int Length { get; }
    /// <summary>The most specific recognized category.</summary>
    public GameEventScriptSyntaxKind Kind { get; }
    /// <summary>Immutable outer-to-inner TextMate scopes, including the root language scope.</summary>
    public IReadOnlyList<string> Scopes { get; }

    internal GameEventScriptHighlightSpan(int start, int length, string[] scopes)
    {
        Start = start;
        Length = length;
        Scopes = Array.AsReadOnly(scopes);
        Kind = ScopeKinds.Classify(scopes);
    }
}

/// <summary>An immutable, equatable line-boundary state. It is not a serializable editor document or compiler state.</summary>
public sealed class GameEventScriptHighlightState : IEquatable<GameEventScriptHighlightState>
{
    /// <summary>The language this state belongs to; states cannot be reused for another language.</summary>
    public GameEventScriptSyntaxLanguage Language { get; }
    internal int[] Frames { get; }

    internal GameEventScriptHighlightState(GameEventScriptSyntaxLanguage language, IEnumerable<int> frames)
    {
        Language = language;
        Frames = frames.ToArray();
    }

    /// <summary>Compares grammar context for incremental retokenization convergence.</summary>
    public bool Equals(GameEventScriptHighlightState? other) => other is not null && Language == other.Language && Frames.SequenceEqual(other.Frames);
    /// <summary>Compares grammar context for incremental retokenization convergence.</summary>
    public override bool Equals(object? obj) => obj is GameEventScriptHighlightState other && Equals(other);
    /// <summary>Returns a hash consistent with grammar-state equality.</summary>
    public override int GetHashCode()
    {
        var hash = (int)Language;
        foreach (var frame in Frames) hash = unchecked(hash * 31 + frame);
        return hash;
    }
}

/// <summary>A highlighting result. Failed resource bounds return no spans or usable continuation state.</summary>
public sealed class GameEventScriptHighlightResult
{
    /// <summary>Ordered, nonoverlapping ranges covering the complete input when successful.</summary>
    public IReadOnlyList<GameEventScriptHighlightSpan> Spans { get; }
    /// <summary>Whether all input was highlighted within the resource limits.</summary>
    public bool IsComplete { get; }
    /// <summary>Context after the final line, or null when incomplete. Do not cache an incomplete result.</summary>
    public GameEventScriptHighlightState? NextState { get; }

    internal GameEventScriptHighlightResult(List<GameEventScriptHighlightSpan> spans, GameEventScriptHighlightState? state)
    {
        Spans = spans.AsReadOnly();
        NextState = state;
        IsComplete = state is not null;
    }
}

internal static class ScopeKinds
{
    internal static GameEventScriptSyntaxKind Classify(IReadOnlyList<string> scopes)
    {
        for (var i = scopes.Count - 1; i >= 0; i--)
        {
            var scope = scopes[i];
            bool Is(string prefix) => scope.StartsWith(prefix, StringComparison.Ordinal);
            if (Is("comment")) return GameEventScriptSyntaxKind.Comment;
            if (Is("string") || Is("constant.character.escape")) return GameEventScriptSyntaxKind.String;
            if (Is("constant.numeric") || Is("constant.language.numeric")) return GameEventScriptSyntaxKind.Number;
            if (Is("variable.other.constant")) return GameEventScriptSyntaxKind.Constant;
            if (Is("variable.other.register")) return GameEventScriptSyntaxKind.Register;
            if (Is("variable.other.label") || Is("constant.other.symbol")) return GameEventScriptSyntaxKind.Label;
            if (Is("entity.name.type.class.message") || Is("entity.name.function.handler")) return GameEventScriptSyntaxKind.Message;
            if (Is("support.type") || Is("entity.name.type")) return GameEventScriptSyntaxKind.Type;
            if (Is("entity.name.namespace")) return GameEventScriptSyntaxKind.Module;
            if (Is("entity.name.function")) return GameEventScriptSyntaxKind.Function;
            if (Is("entity.other.attribute-name.tag")) return GameEventScriptSyntaxKind.Tag;
            if (Is("support") || Is("entity")) return GameEventScriptSyntaxKind.Builtin;
            if (Is("keyword") || Is("constant.language.boolean")) return GameEventScriptSyntaxKind.Keyword;
            if (Is("variable")) return GameEventScriptSyntaxKind.Identifier;
            if (Is("punctuation")) return GameEventScriptSyntaxKind.Symbol;
        }
        return GameEventScriptSyntaxKind.Plain;
    }
}
